### PR TITLE
Add Cirrus CI configuration to test capnproto on a few actively maintained FreeBSD versions and fix few issues

### PR DESCRIPTION
**Pull request by sobomax**
This should run automatically on PRs and commits and will ensure no regressions are made over time (might need project maintainers to make few clicks to enable).

Few issues preventing unit test suite from completing have been identified and fixed.

### CODE REVIEW

File: c++/src/kj/string-test.c++
Comment (sobomax) [reply c++/src/kj/string-test.c++:286]: OK, I've added es_ES.utf8 back. Thanks!

File: c++/src/kj/async.c++
Comment (kentonv) [inline c++/src/kj/async.c++:1316]: Looks like the meaning of the check changed here? `_WIN32` and `__CYGWIN__` should avoid this block.

File: c++/src/kj/async.c++
Comment (kentonv) [inline c++/src/kj/async.c++:1474]: I can't figure out what `#if` this is matching. Maybe time to add comments to the `#endif`s to make clear? Like:

```
#endif  // KJ_USE_FIBERS
```

Comment (kentonv) [reply c++/src/kj/async.c++:1474]: Oh I see it's for KJ_NO_EXCEPTIONS. Perhaps we should roll that into `KJ_USE_FIBERS`? I.e. make sure to define `KJ_USE_FIBERS 0` when `KJ_NO_EXCEPTIONS` is enabled.

File: c++/src/kj/async-test.c++
Comment (kentonv) [inline c++/src/kj/async-test.c++:30]: Can we just remove this define entirely and replace `KJ_FIBERS_AVAILABLE` with `KJ_USE_FIBERS`? (See the TODO, above.)

File: c++/src/kj/async-io-test.c++
Comment (kentonv) [inline c++/src/kj/async-io-test.c++:306]: Nitpick, but I feel that the negation-of-ors is harder to follow than the previous ands-of-negations, can we keep it the old way and just add freeBSD to the end?

File: c++/src/kj/async.c++
Comment (sobomax) [reply c++/src/kj/async.c++:1316]: Fixed.

File: c++/src/kj/async.c++
Comment (sobomax) [reply c++/src/kj/async.c++]: Done. It's bit hairy now that you start considering all permutations of what might be defined and what not.... But let's see if you like it and go from there. )

File: c++/src/kj/async-test.c++
Comment (sobomax) [reply c++/src/kj/async-test.c++]: Done.

File: c++/src/kj/async.c++
Comment (sobomax) [reply c++/src/kj/async.c++]: Done. See my comment on that section becoming quite hairy. :)

File: c++/src/kj/async-io-test.c++
Comment (sobomax) [reply c++/src/kj/async-io-test.c++]: Done. Your version is 3 character longer. :-P

```
-#if !(_WIN32 || __CYGWIN__  || __APPLE__ || __ANDROID__ || __FreeBSD__)
+#if !_WIN32 && !__CYGWIN__  && !__APPLE__ && !__ANDROID__ && !__FreeBSD__
```

**Comment by sobomax** [comment]
> This looks very promising, especially the CI. For the Fibers, could you move the fibers to be using a `KJ_USE_FIBERS` macro so that we don't have this logic duplicated all over as discussed in #1313, #1208, & #1167?

Done, thanks!

File: c++/src/kj/async-test.c++
Comment (sobomax) [inline c++/src/kj/async-test.c++:29]: @vlovich This comment seems to be somewhat stale now. Not sure if I just need to remove it?

Comment (kentonv) [reply c++/src/kj/async-test.c++:29]: Yeah, I'd remove this.

File: c++/src/kj/async.c++
Comment (kentonv) [inline c++/src/kj/async.c++:1468]: Can we bring back the custom error message for KJ_NO_EXCEPTIONS:

```
#if KJ_NO_EXCEPTIONS
    KJ_UNIMPLEMENTED("Fibers are not implemented because exceptions are disabled");
#else
    KJ_UNIMPLEMENTED(
      "Fibers are not implemented on this platform because its C library lacks setcontext() "
      "and friends. If you'd like to see fiber support added, file a bug to let us know. "
      "We can likely make it happen using assembly, but didn't want to try unless it was "
      "actually needed.");
#endif
```

File: c++/src/kj/async.h
Comment (kentonv) [inline c++/src/kj/async.h:45]: Let's use indentation here, and simplify things a bit. There's no need to check if KJ_NO_EXCEPTIONS is defined; if it's not defined, the preprocessor treats it as falsy. We generally use `#if KJ_NO_EXCEPTIONS` without first checking if it's defined elsewhere in KJ code.

```suggestion
#ifndef KJ_USE_FIBERS
  #if __BIONIC__ || __FreeBSD__ || __OpenBSD__ || KJ_NO_EXCEPTIONS
    // These platforms don't support fibers.
    #define KJ_USE_FIBERS 0
  #else
    #define KJ_USE_FIBERS 1
  #endif
#else
  #if KJ_NO_EXCEPTIONS && KJ_USE_FIBERS
    #error "Fibers cannot be enabled when exceptions are disabled."
  #endif
#endif
```

File: c++/src/kj/async-io-test.c++
Comment (kentonv) [inline c++/src/kj/async-io-test.c++:306]: Actually looking at this again, if we're literally filtering every platform except linux, maybe we should just make it `#if __linux__ && !__ANDROID__`?

```suggestion
// SO_TIMESTAMP specifically seems to have lots of portability issues, so we only even attempt this
// test on Linux. Even Android fails this test for some unknown reason, so we avoid it there. This
// is an extremely obscure feature and it's just not worth the time to debug the test on every
// platform.
#if __linux__ && !__ANDROID__
```

File: .cirrus.yml
Comment (kentonv) [inline .cirrus.yml:1]: Is there really no way to run FreeBSD on GitHub Actions? I was pretty happy that we managed to consolidate our CI a while back...

Comment (sobomax) [reply .cirrus.yml:1]: Not without major hassle, unfortunately. Like via quemu or something of that sort. But in my experience using Cirrus with other projects it has quite low overhead and is otherwise "set and forget". Examples of some other projects that use both Actions and Cirrus are git itself: https://github.com/git/git/runs/3416539724

File: c++/src/kj/async-io-test.c++
Comment (sobomax) [reply c++/src/kj/async-io-test.c++:306]: That comment is somewhat misleading in my view, I must point out. There is nothing wrong with SO_TIMESTAMP support (it's been around way before Linux came along), the only problem is that the test case specifically written around using that feature with SOCK_STREAM, a (mis-)feature that has been added into Linux kernel around 2009. I even dug a commit message at some point, but is too lazy to find it out. The whole idea of timestamping a data on a stream connection has no much sense, since you might be reading data from multiple TCP packets in the same recvmsg() call.

Comment (sobomax) [reply c++/src/kj/async-io-test.c++:306]: P.S. The correct way to fix this would be to re-write test case to use datagram sockets, IMHO. (not volunteering) :)

File: c++/src/kj/async.c++
Comment (sobomax) [reply c++/src/kj/async.c++:1471]: Done. I am wondering if it's d3adc0d3() now with an #error statement in the async.h?

File: c++/src/kj/async-test.c++
Comment (sobomax) [reply c++/src/kj/async-test.c++]: Removed.

File: c++/src/kj/async-io-test.c++
Comment (kentonv) [reply c++/src/kj/async-io-test.c++:306]: Sure, I didn't write this test and wasn't aware of the specific details. I believe the test is specifically trying to test ancillary messages on streams, though, so changing it to test datagram sockets would defeat the purpose. How about:

```
// This test uses SO_TIMESTAMP on a SOCK_STREAM, which is only supported by Linux. Ideally we'd
// rewrite the test to use some other message type that is widely supported on streams. But for
// now we just limit the test to Linux. Also, it doesn't work on Android for some reason, and it
// isn't worth investigating, so we skip it there.
```

File: c++/src/kj/async.c++
Comment (kentonv) [reply c++/src/kj/async.c++:1471]: No... that error will only happen if `KJ_USE_FIBERS=1`, but this code here is for `KJ_USE_FIBERS=0`.

**Review by kentonv — APPROVED** [review]
This looks good but I can't merge as it's still marked "draft".

**Comment by sobomax** [comment]
> This looks good but I can't merge as it's still marked "draft".

Give me few minutes, I am polishing few things. The last forced push is still pending, I'll move it into ready state once it's in.

**Comment by kentonv** [comment]
Unfortunately GitHub doesn't provide diffs across force-pushes so I'd actually prefer that you use incremental commits and I can squash-merge later. Otherwise I have to review the whole thing every time...

**Comment by sobomax** [comment]
> Unfortunately GitHub doesn't provide diffs across force-pushes so I'd actually prefer that you use incremental commits and I can squash-merge later. Otherwise I have to review the whole thing every time...

Oh, ok, will keep it in mind. You never know. I remember some of the other maintainers have been upset by me having 10-20 commits in the PR and demanding me to do squashing. So it's kinda habitual now...

**Comment by sobomax** [comment]
@kentonv I can actually see the diff between forced requests (b600c9a was immediately before approval, 7a82d79 is the very last one). Not sure if it would work for code pulled from the github:

```
$ git diff   b600c9a..7a82d79
diff --git a/c++/src/kj/async-io-test.c++ b/c++/src/kj/async-io-test.c++
index bc833716..e2d12401 100644
--- a/c++/src/kj/async-io-test.c++
+++ b/c++/src/kj/async-io-test.c++
@@ -299,11 +299,11 @@ TEST(AsyncIo, AncillaryMessageHandlerNoMsg) {
 }
 #endif

-// We only support ancillary messages on Unix.
-// MacOS & FreeBSD only support SO_TIMESTAMP on datagram sockets, so this test won't work.
-// Android fails this test for unknown reasons and I don't care because it's an extremely obscure
-// feature anyway.
-#if !_WIN32 && !__CYGWIN__  && !__APPLE__ && !__ANDROID__ && !__FreeBSD__
+// This test uses SO_TIMESTAMP on a SOCK_STREAM, which is only supported by Linux. Ideally we'd
+// rewrite the test to use some other message type that is widely supported on streams. But for
+// now we just limit the test to Linux. Also, it doesn't work on Android for some reason, and it
+// isn't worth investigating, so we skip it there.
+#if __linux__ && !__ANDROID__
 TEST(AsyncIo, AncillaryMessageHandler) {
   auto ioContext = setupAsyncIo();
   auto& network = ioContext.provider->getNetwork();
$

File: .cirrus.yml
Comment (kentonv) [inline .cirrus.yml:30]: Hey sorry, I should have looked more closely at this earlier. Can you use `super-test.sh` like all the other Unix CI targets, rather than building manually? It's specifically intended to be our CI script.

Comment (kentonv) [reply .cirrus.yml:30]: Though if it's easier to merge this now and get CI configured before iterating on that, I'm OK with that... but we should have this use `super-test.sh` long-term.

Comment (sobomax) [reply .cirrus.yml:30]: @kentonv I gave it a brief try and it looks like super-test.sh has some bash-only features and also fails to detect system Clang for some reason on FreeBSD. Therefore, I suggest perhaps flushing this PR since it affects other PRs in pipeline and I'll open a new one to switch Cirus to use super-test.sh once I have a chance to massage it to complete on FreeBSD. Thanks!

**Comment by sobomax** [comment]
Hi @kentonv any chance to get this merged today, so that I can start massaging super-test.sh to make it usable on FreeBSD/Cyrus? I've got some interesting new failures coming out of it, having a working CI in master would be a good help to test against baseline.

```
ld: error: undefined symbol: kj::InFlightExceptionIterator::InFlightExceptionIterator()
>>> referenced by exception-test.c++
>>>               src/kj/capnp_test-exception-test.o:(kj::_::(anonymous namespace)::TestCase212::run())

ld: error: undefined symbol: kj::InFlightExceptionIterator::next()
>>> referenced by exception-test.c++
>>>               src/kj/capnp_test-exception-test.o:(kj::_::(anonymous namespace)::TestCase212::run())
>>> referenced by exception-test.c++
>>>               src/kj/capnp_test-exception-test.o:(kj::_::(anonymous namespace)::TestCase212::run())
>>> referenced by exception-test.c++
>>>               src/kj/capnp_test-exception-test.o:(kj::_::(anonymous namespace)::TestCase212::run())
c++: error: linker command failed with exit code 1 (use -v to see invocation)
*** [capnp-test] Error code 1

**Comment by kentonv** [comment]
Thanks for working on this!

Note that I only do code reviews M/W/F, and I have dozens of code reviews every week, so I may be a bit slow at times. Sorry.

**Comment by sobomax** [comment]
No worries, @kentonv thanks for taking care of this. Could you also look into enabling Cirrus CI for the capnproto project, so that it would integrate FreeBSD runs into PRs and commits? You can find instructions here: https://cirrus-ci.org/guide/quick-start/

I am on to get super-test running there.

**Comment by kentonv** [comment]
Hi @sobomax,

I've put together a point release including the FreeBSD changes:

https://capnproto.org/capnproto-c++-0.9.1-rc1.tar.gz

Can you verify it works?

Sorry I haven't gotten around to setting up the CI yet.

**Pull request merged**
